### DISCUSSION TITLE
Circuit box xml sometimes has incorrect usedresource for components

### DISCUSSION THREAD

**Discussion by Max112358** [FakeFishGames/Barotrauma#14388] [Bug reports]
### Disclaimers

- [X] I have searched the issue tracker to check if the issue has already been reported.
- [X] My issue happened while using mods.

### What happened?

I created the mod Blueprints on the steam workshop. It saves and loads circuitboxes based on their XML data. 

Sometimes circuitboxes XML data will change from the base resource used to input each component (such as "memorycomponent" into "fpgacircuit". This messes my Blueprints mod up because that's the info I use to rebuild the circuit. 

This only happens sometimes and I'm not sure what causes it. Even more confusingly I've seen circuitboxes in the wild on multiplayer that have only some components this happened to.

This is a link to the mod if you want it for testing purposes.
https://steamcommunity.com/sharedfiles/filedetails/?id=3301014847

[absurdly_complex_reactor.txt](https://github.com/user-attachments/files/16463829/absurdly_complex_reactor.txt)
[default_docking_hatch.txt](https://github.com/user-attachments/files/16463830/default_docking_hatch.txt)

![xml_bug](https://github.com/user-attachments/assets/6695456c-86f5-49b8-816f-af79ddde0406)

### Reproduction steps

Start a multiplayer campaign.
Make a circuit box. Attach it to a wall and make a circuit inside it.
Observe the XML data of the box. You can get this via the item.save() function built into items.
BreadcrumbsBarotrauma/Barotrauma/BarotraumaShared/SharedSource/Items/Item.cs
You can get it easily by using my Blueprints mod and just saving the circuitbox to a file (all it does is run this function)
See that the circuitbox is normal.
Play the game for a while.
Check again, the XML has changed to have the "fpgacircuit".

If you just want to see one, the dugong docking hatch box seems to have this bug from the start of the game.

### Bug prevalence

Happens every now and then

### Single player or multiplayer?

Multiplayer hosted from the in-game menu (= using a listen server)

### Version

v1.5.9.1 (Summer Update Hotfix 2)

### Which operating system did you encounter this bug on?

Windows

### Relevant error messages and crash reports

```shell
No errors or crash reports.
```

**Comment by 3e849f2e5c** [comment]
I think there might be a misunderstanding on what the `usedresource` is.

For convenience the circuit box allows you to use both the component itself or a FPGA circuit to attach a component inside of it, that way you don't have to go out of your way to craft each component while also allowing you to spend your surplus components without deconstructing them.
![Barotrauma_jWuDkMg1K8](https://github.com/user-attachments/assets/7a937760-a45a-40a5-aa32-2d8edcf3813d)


`usedresource` is used by the game to track what resource you used to attach the component so it can be refunded to the player when you detach the component.

When working with circuit boxes in sub editor the `usedresource` will always be a FPGA circuit since placing them is free in sub editor and it was easier to always default to FPGA circuit for everything in a case we don't consume any resources.

Unless I'm misunderstanding something I don't think `usedresource` is incorrectly being reverted to an incorrect item here

**Reply by Max112358** [reply]
Well that explains some things. If that is the case then I would humbly request that the XML somewhere contains what kind of components the circuitbox holds, just so it can be recreated later.

Something like:
`<Component id="6" position="1119.8271,1105.6787" backingitemid="6,0" usedresource="fpgacircuit" item="multiplycomponent" /> `

Basically just add the item to the component section.

Also as long as I'm asking for stuff, it would be great if there was a function to make IReadOnlyCollection<CircuitBoxNode> out of a CircuitBoxNode. So just take a circuitboxnode as a parameter, and return an IReadOnlyCollection<CircuitBoxNode> of that node. Probably more generic than that though, so any element and make an IReadOnlyCollection out of it.
